### ISSUE TITLE
Conver To > PDE Project does not add the java nature

### ISSUE THREAD

**Issue by laeubi** [eclipse-pde/eclipse.pde#338]
Currently on can convert a project to a PDE project, but if this is not already a java project this does not work well because the java nature is not added.

**Comment by mickaelistria** [comment]
Not all PDE project have to be Java project. Eg in Platform there is `org.eclipse.ui.themes` that's not a Java project. So I'm unsure this is generally desirable to enable Java nature by default. Maybe the action could verify that the project contains some .java file first.

**Comment by laeubi** [comment]
But whats the purpose of it being a "pde project" if it does not contain any java code? Would then be a plain project be sufficient?

**Comment by mickaelistria** [comment]
A PDE project enables
* Dependency resolution based on MANIFEST.MF
* Run As > Eclipse application... launchers on the projects
* Detection by the other launch configuration, inclusion in launches applications
* Detection by other possible PDE integration stuff, such as API analysis (which might proceed things like plugin.xml)
* ...

So it's (unfortunately) not an empty decoration but something many other tools rely on. It would be nicer if those tools where ignoring the nature and were only looking for a suitable MANIFEST.MF with some Bundle-SymbolicName, but there is a lot of legacy to change...

**Comment by laeubi** [comment]
> * Dependency resolution based on MANIFEST.MF
> Detection by the other launch configuration, inclusion in launches applications
> * Run As > Eclipse application... launchers on the projects

Does not sounds very useful if there is no code, maybe for a fragment that only contains resources, but would the java nature harm anything?

> * Detection by other possible PDE integration stuff, such as API analysis (which might proceed things like plugin.xml)

API analysis without any java code?

> So it's (unfortunately) not an empty decoration but something many other tools rely on.

And I assume most if not all check for the java nature first, at least that the usual PDE code and for almost if not all cases a PDE nature implies a java nature ...

**Comment by mickaelistria** [comment]
> Does not sounds very useful if there is no code

that has been multiple times useful to me, and I guess to others.

> maybe for a fragment that only contains resources, but would the java nature harm anything?

Maybe, but still, this case of a bundle without Java code is a valid one.

> API analysis without any java code?

As mentioned, API analysis is kind of expected (not sure it does though) to also verify plugin.xml

> nd I assume most if not all check for the java nature 

I just gave you several counter-examples that not all check the java nature, so you assume wrong.

**Comment by HannesWell** [comment]
> > * Dependency resolution based on MANIFEST.MF
> >   Detection by the other launch configuration, inclusion in launches applications
> > * Run As > Eclipse application... launchers on the projects
> 
> Does not sounds very useful if there is no code, maybe for a fragment that only contains resources, but would the java nature harm anything?

Probably the only harm is that if the java-nature is present a missing .classpath file is indicated as an error. I haven't verified it, just guessing it, but if that isn't an error/warning, then I don't see any other harm.

**Comment by laeubi** [comment]
> that has been multiple times useful to me, and I guess to others.

If on absolutely needs and is certain he can remove the java nature, but having a PDE project without a java nature is really a very seldom edge case and generally not useful.

> I just gave you several counter-examples that not all check the java nature, so you assume wrong.

These are not real "counter examples" ... you are just listing some very rare cases where one can trigger some functionality, but none of those are "common cases" nor do they do not work or get harmed anything if the java nature would have been enabled.

On the other side, **not enabling** the java nature makes the whole "convert to PDE Project" useless for most cases where one wants to convert a project to a plugin ....

**Comment by laeubi** [comment]
> Probably the only harm is that if the java-nature is present a missing .classpath file is indicated as an error. I haven't verified it, just guessing it, but if that isn't an error/warning, then I don't see any other harm.

A classpath file can be completely empty (if one wants it to be ...)

**Comment by laeubi** [comment]
> * Dependency resolution based on MANIFEST.MF

Just out of curiosity I tried this out, a project without java nature/builder has no PDE Dependencies resolved.

**Comment by mickaelistria** [comment]
> These are not real "counter examples" ...

Well, maybe "counter exmaple" translate to something different in both our native language, but in mine, a counter-example is whatever example for which a general assumption is not true; and it's sufficient to use in order to prove the general assumption is wrong.

> Just out of curiosity I tried this out, a project without java nature/builder has no PDE Dependencies resolved.

It has dependency resolved if you include it in some launch configuration and click "Add required plugins", or "validate". It's not visible in the Project Explorer because the Java model is not affected, but the PDE one is.

**Comment by HannesWell** [comment]
> > Probably the only harm is that if the java-nature is present a missing .classpath file is indicated as an error. I haven't verified it, just guessing it, but if that isn't an error/warning, then I don't see any other harm.
> 
> A classpath file can be completely empty (if one wants it to be ...)

That's right. Personally I don't like to keep around empty files, but in the rare case of resource-only plug-ins one can simply remove the .classpath file and the java-nature manually. So adding it by default seem reasonable for me, since in most cases it is desired.

**Comment by laeubi** [comment]
> It has dependency resolved if you include it in some launch configuration and click "Add required plugins", or "validate"

But this function do not use the "resolved project dependencies" so it is not only not visible it is simply not used and not resolved for the project.

> Well, maybe "counter exmaple" translate to something different in both our native language, but in mine, a counter-example is whatever example for which a general assumption is not true;

Well then we have at least one counter-example that **not** enabling it is also wrong, so the question is:

Should we not better then cover the common case than the seldom case when there is no general assumption that is true for every case?

**Comment by mickaelistria** [comment]
> Should we not better then cover the common case than the seldom case when there is no general assumption that is true for every case?

The action can make the distinction and detect whether or not the folder has some Java files. Why not just making the conversion action smarter instead of adding constraints on unrelated well-formed and stable projects?

**Comment by laeubi** [comment]
> Why not just making the conversion action smarter instead of adding constraints on unrelated well-formed and stable projects?

1. I have never stated any solution just a requirement, that is that converting a project to PDE should add the java nature (probably detecting some other stuff
2. As this is a manual action you can only trigger once for a project, this won't have any influence on existing "well-formed and stable projects" anyways ...

So this discussion is a bit strange to me because the current behavior produces hard to fix state in many cases, while the proposed solution only adds a little overhead that do not harm anyway and it is generally easier to remove the builder I don't need then finding out how it is named and adding it manually for the 95% cases ...

**Comment by laeubi** [comment]
And just for completeness if I **create** a PDE Plugin project it always adds the java nature... so why should converting be different than creating it from scratch?

**Comment by mickaelistria** [comment]
> And just for completeness if I create a PDE Plugin project it always adds the java nature...

Not always, there is an option to explicitly create it (this option is on by default).

**Comment by merks** [comment]
Why not make it an option like in the New Plug-in Project Wizard?  An option that is only present if it's not already a Java project. But I have even more general questions. Why is there nothing to convert a plain project to a Java project?   Why did someone create a plain project that is not a Java project and then later decide it should be a both Plug-in project and also a Java project.  That seems kind of far-fetched to me.  But surely if that is not far fetched, then some (more?) people also need to convert a plain project to a Java project while that functionality is completely missing.  If it were not, what's described in this issue wouldn't be needed because it could make use of that.
